File: 1250-longest-common-subsequence/longest-common-subsequence.cpp
class Solution {
public:
    int solve(string &text1, string &text2, int n, int m, vector<vector<int>> &memo){
        if (n == 0 || m == 0) return 0;
        if (memo[n][m] != -1) return memo[n][m];

        if (text1[n-1] == text2[m-1]){
            return memo[n][m] = 1 + solve(text1, text2, n-1, m-1, memo);
        } else {
            return memo[n][m] = max(solve(text1, text2, n-1, m, memo), solve(text1, text2, n, m-1, memo));
        }
    }
    int longestCommonSubsequence(string text1, string text2) {
        vector<vector<int>> dp(text1.length()+1, vector<int> (text2.length()+1, -1));
        return solve(text1, text2, text1.length(), text2.length(), dp);
    }
};

File: 1250-longest-common-subsequence/longest-common-subsequence.go
func solve(s, t string, i, j int, dp [][]int) int {
    if i >= len(s) || j >= len(t) {
        return 0
    }

    if dp[i][j] != -1 {
        return dp[i][j]
    }

    if s[i] == t[j] {
        dp[i][j] = 1 + solve(s, t, i + 1, j + 1, dp)
    } else {
        dp[i][j] = max(solve(s, t, i + 1, j, dp), solve(s, t, i, j + 1, dp))
    }
    return dp[i][j]
}

func longestCommonSubsequence(text1 string, text2 string) int {
    n, m := len(text1), len(text2)
    dp := make([][]int, n)
    for i := range dp {
        dp[i] = make([]int, m)
        for j := range dp[i] {
            dp[i][j] = -1
        }
    }
    return solve(text1, text2, 0, 0, dp)
}